packages/web-sdk/src/instrumentations/userActions/userActionController.ts

```diff
@@ -0,0 +1,146 @@
+    this.allMonitorsSub = new Observable()
+      .merge(this.http, this.dom, this.perf)
+      .takeWhile(() => [UserActionState.Started, UserActionState.Halted].includes(this.ua.getState()))
+      .filter((msg) => {
+        // If the user action is in halt state, we only keep listening to ended http requests
+        if (
+          this.ua.getState() === UserActionState.Halted &&
+          !(isRequestEndMessage(msg) && this.runningRequests.has(msg.request.requestId))
+        ) {
+          return false;
+        }
+
+        return true;
+      })
+      .subscribe((msg) => {
+        if (isRequestStartMessage(msg)) {
+          // An action is on halt if it has pending items, like pending HTTP requests.
+          // In this case we start a separate timeout to wait for the requests to finish
+          // If in the halt state, we stop adding Faro signals to the action's buffer (see userActionLifecycleHandler.ts)
+          // But we are still subscribed to
+          this.runningRequests.set(msg.request.requestId, msg.request);
+        }
+
+        if (isRequestEndMessage(msg)) {
+          this.runningRequests.delete(msg.request.requestId);
+        }
+
+        if (!isRequestEndMessage(msg)) {
+          if (!this.isValid) {
+            this.isValid = true;
+          }
+          this.scheduleFollowUp();
+        } else if (this.isHalted && this.runningRequests.size === 0) {
+          this.endAction();
+        }
+      });
+
+    // When UA ends or cancels, cleanup timers/subscriptions
+    this.stateSub = (this.ua as unknown as Observable)
+      .filter((s: UserActionState) => [UserActionState.Ended, UserActionState.Cancelled].includes(s))
+      .first()
+      .subscribe(() => this.cleanup());
+
+    // initial follow-up window in case nothing else happens
+    this.scheduleFollowUp();
+  }
+
+  private scheduleFollowUp() {
+    this.clearTimer(this.followUpTid);
+    this.followUpTid = setTimeout(() => {
+      // If action just started and there's pending work, go to halted
+      if (this.ua.getState() === UserActionState.Started && this.runningRequests.size > 0) {
+        this.haltAction();
+        return;
+      }
+
+      // If we saw any relevant activity in the window, finish as ended
+      if (this.isValid) {
+        this.endAction();
+        return;
+      }
+
+      // Otherwise, no signals => cancel
+      this.cancelAction();
+    }, defaultFollowUpActionTimeRange) as any;
+  }
+
+  private haltAction() {
+    if (this.ua.getState() !== UserActionState.Started) {
+      return;
+    }
+    this.isHalted = true;
+    this.startHaltTimeout();
+  }
+
+  private startHaltTimeout() {
+    this.clearTimer(this.haltTid);
+    this.haltTid = startTimeout(
+      this.haltTid,
+      () => {
+        // If still halted after timeout, end
+        if (this.ua.getState() === UserActionState.Halted) {
+          this.endAction();
+        }
+      },
+      defaultHaltTimeout
```

Comment:
This used to be a configurable parameter. I would suggest modifying the API a little bit so that users can still make use of it.